apps/next/app/login/page.tsx

```diff
@@ -1,16 +1,27 @@
 import { redirect } from "next/navigation"
 
+import { SearchParamError } from "@/lib/types/error"
 import { LoginForm } from "@/components/user/login-form"
 
 import { createClient } from "@/modules/utils/server"
 
-export default async function Page() {
+export default async function Page({
+  searchParams,
+}: {
+  searchParams: { error?: string }
+}) {
   const supabase = createClient()
 
   const {
     data: { user },
   } = await supabase.auth.getUser()
 
+  if (searchParams.error) {
+    return (
+      <LoginForm error={JSON.parse(searchParams.error) as SearchParamError} />
```

Comment:
Can we use Zod to parse this, make sure to use a catch chaining with a sensible fallback and skip the type casting?

Also can we write the redirects above the components? Here we have component, redirect, and component.

Reply:
Thanks for the tip. I am quite new to the zod library but I've studied docs and motivation behind it and hopefully I finally understood it.